### PR TITLE
Add content shell support to wptrunner.

### PR DESCRIPTION
**Pull request by t-troebst**
This PR adds support for the Chromium content_shell in wptrunner. content_shell represents a minimal implementation of the  content module and is what is currently tested against in the Chromium CI pipeline. We would like to start running wptrunner with content_shell in the pipeline as well.

Tests for content_shell are normally run through a very rudimentary protocol (see [here](https://chromium.googlesource.com/chromium/src.git/+/HEAD/content/web_test/browser/test_info_extractor.h)) that uses stdout / stdin. The ContentShellBrowser relays the information using multiprocessing.Queues so that the executors can send commands to the process.

Currently, this supports ref tests, crash tests, and testharness.js tests (with the exception of testdriver.js).

### CODE REVIEW

**Comment by t-troebst** [comment]
This change underwent some previous review in the Chromium project ([here](https://chromium-review.googlesource.com/c/chromium/src/+/3788909)) with @WeizhongX @jonathan-j-lee.

File: tools/wptrunner/wptrunner/executors/executorcontentshell.py
Comment (jgraham) [inline tools/wptrunner/wptrunner/executors/executorcontentshell.py:71]: Hmm, this seems like it's also affecting state. Really `is_alive` should be safe to call at any time. In this case maybe it makes sense to check that the queues are still open, or similar.

File: tools/wptrunner/wptrunner/browsers/content_shell.py
Comment (jgraham) [inline tools/wptrunner/wptrunner/browsers/content_shell.py:15]: Try to avoid using line continulation characters, prefer using parens for longer expressions.

File: tools/wptrunner/wptrunner/executors/executorcontentshell.py
Comment (jgraham) [inline tools/wptrunner/wptrunner/executors/executorcontentshell.py:77]: So it seems unusual that this doesn't implement the `Protocol` API but defines a totally new interface that's specific to ContentShell. I see that the protocol is quite different from the usual WebDriver-esque protocols, but I think it would be better to conform to the existing API to the extent possible.

**Review by jgraham — COMMENTED** [review]
Generally this looks pretty good, thanks. There are a few style comments, but the main substantive thing is that the Protocol implementation is defining a totally novel API that's unrelated to the class it's inheriting from.

File: tools/wptrunner/wptrunner/browsers/content_shell.py
Comment (t-troebst) [reply tools/wptrunner/wptrunner/browsers/content_shell.py]: I've added a timeout to the `join()` calls for the IO threads. I think `multiprocessing.Queue.join_thread` doesn't accept a timeout though I don't think there is too much of a risk of it hanging.

Comment (t-troebst) [reply tools/wptrunner/wptrunner/browsers/content_shell.py]: I ended up further moving the join_thread calls into the threads.

File: tools/wptrunner/wptrunner/browsers/content_shell.py
Comment (t-troebst) [reply tools/wptrunner/wptrunner/browsers/content_shell.py]: Done.

File: tools/wptrunner/wptrunner/browsers/content_shell.py
Comment (t-troebst) [reply tools/wptrunner/wptrunner/browsers/content_shell.py:36]: Done.

File: tools/wptrunner/wptrunner/browsers/content_shell.py
Comment (t-troebst) [reply tools/wptrunner/wptrunner/browsers/content_shell.py]: Done.

File: tools/wptrunner/wptrunner/executors/executorcontentshell.py
Comment (t-troebst) [reply tools/wptrunner/wptrunner/executors/executorcontentshell.py]: I've changed it so that it checks if the IO pipes are still open. The problem with this is that we cannot know if the process is still responsive. Unfortunately, I don't know of a way to check that for content_shell without affecting state.

File: tools/wptrunner/wptrunner/browsers/content_shell.py
Comment (t-troebst) [reply tools/wptrunner/wptrunner/browsers/content_shell.py:94]: Yes, this is very similar to how mozprocess handles output with two distinctions.
1. We use `multiprocessing.Queue` so that other processes can interact with the browser through its ExecutorBrowser. In mozprocess, the IO has to happen in the same process.
2. mozprocess calls `rstrip()` on the lines it gets from stdout before passing it to the output handler. We don't want that, especially when reading binary image data.

File: tools/wptrunner/wptrunner/executors/executorcontentshell.py
Comment (t-troebst) [reply tools/wptrunner/wptrunner/executors/executorcontentshell.py:75]: I don't see how we could conform to the `Protocol` API basically at all. The only thing that the content_shell testing protocol supports right now is sending a single URL and getting text, image, or audio output. We can run the tests but we cannot implement any of the `ProtocolPart` classes. So the custom logic that is shared between `Executor` classes has to go somewhere. I see four options:
1. Put the logic in the `Protocol` class.
2. Create a new `ProtocolPart` and put the logic in there.
3. Put the logic in a custom `ExecutorBrowser` class.
4. Put the logic in a completely separate class / functions.

From what I can see in the codebase, [executorselenium.py](https://github.com/web-platform-tests/wpt/blob/0d6d9dc80bf23f94fab8137920b73228690b6de7/tools/wptrunner/wptrunner/executors/executorselenium.py#L267) does (essentially) option 1 by adding a `webdriver` attribute to its `Protocol` and using that for custom logic in some places. [executorservo.py](https://github.com/web-platform-tests/wpt/blob/0d6d9dc80bf23f94fab8137920b73228690b6de7/tools/wptrunner/wptrunner/executors/executorservo.py#L72) does option 4 and just uses an empty `ConnectionlessProtocol` (which is what we inherit from).

**Comment by WeizhongX** [comment]
@foolip, @jgraham, a friendly ping...

**Comment by DanielRyanSmith** [comment]
It looks like some mypy errors exist already in the repo and are causing unrelated failures in the  `tools/ unittests` checks. 
```
py310-mypy run-test: commands[0] | mypy --config-file=/home/test/web-platform-tests/tools/mypy.ini tools/
tools/wptrunner/wptrunner/tests/test_metadata.py:9: error: Function is missing a type annotation  [no-untyped-def]
tools/wptrunner/wptrunner/tests/test_metadata.py:19: error: Function is missing a type annotation  [no-untyped-def]
tools/wptrunner/wptrunner/tests/test_metadata.py:36: error: Function is missing a type annotation  [no-untyped-def]
tools/wptrunner/wptrunner/tests/test_metadata.py:42: error: Function is missing a type annotation  [no-untyped-def]
```

**Comment by t-troebst** [comment]
Yes, this is unrelated to this change - just a missing mypy config line in a file that was added recently. I've created a PR to resolve it: https://github.com/web-platform-tests/wpt/pull/35456

File: tools/wpt/browser.py
Comment (jcscottiii) [inline tools/wpt/browser.py:1049]: Hey @t-troebst quick question (I'm fairly new to the project).

For this find_executable, do you want users to have installed content_shell within the virutalenv path?

For reference, when I ran your fork, I had content_shell already in my PATH (because I have my chromium/src/out/Default in my PATH). I had to change this line to `return find_executable("content_shell")` to get it to run. 

Just wanted clarification on how you expect users to have set this up.

Comment (t-troebst) [reply tools/wpt/browser.py:1049]: Hm yes, I was planning to do the installation in a separate PR since this is already getting rather big. We can get the snapshots from the same place that we get the chrome snapshots (with the name "content-shell.zip" instead of "chrome-[platform].zip"). At that point we would install them into the venv and run from there. However, in the mean time I think it is better to look for the executable in the OS path. I'll change that.

Comment (t-troebst) [reply tools/wpt/browser.py]: Oh, and the way that this will run in the Chromium CI is through the "run_wpt_tests.py" script that accepts a "-t" parameter which sets the content_shell binary explicitly.

Comment (jcscottiii) [reply tools/wpt/browser.py]: Thanks for that explanation!

One more question. For those wanting to run locally on windows and mac os, I am reading this [document](https://chromium.googlesource.com/chromium/src/+/HEAD/docs/testing/web_tests_in_content_shell.md#run-web-tests-directly-with-content-shell).

Do we need to do any special case OS matching for `Content Shell.app` for (mac os x)? 

For example: Mac OS
https://github.com/web-platform-tests/wpt/blob/3f6a593433dfad9fe234a616c0dda1bb3deb9625/tools/wpt/browser.py#L953-L957

File: tools/wptrunner/wptrunner/executors/executorcontentshell.py
Comment (jonathan-j-lee) [reply tools/wptrunner/wptrunner/executors/executorcontentshell.py:75]: I have some tentative suggestions:
1. Currently, some details of protocol mode are leaked into `ContentShellBrowser` (e.g., the fact that the protocol is line-based, and that we wait for a `#READY` message). Maybe we could do the following:
    * Reduce `ContentShellBrowser` to managing the subprocess and exposing stdio handles.
    * `ContentShellProtocol` becomes a regular `Protocol` and manages the stdio queues/threads, where `connect()` waits for the `#READY` message.
1. IMO, it's kind of strange that there's a `do_test(...)` method on this protocol (although I guess you could argue this protocol is explicitly designed for testing). There are really two interactions with content shell: running the test and checking for errors on stderr, which I'm leaning towards modeling as separate `ProtocolPart`s (i.e., option 2).

Need to think about this some more.

Comment (t-troebst) [reply tools/wptrunner/wptrunner/executors/executorcontentshell.py:51]: I've done an attempt at this, let me know if this looks better.

File: tools/wpt/browser.py
Comment (t-troebst) [reply tools/wpt/browser.py]: I don't have a Mac device either but I agree that we should probably look for "Content Shell.app/Contents/MacOS/Content Shell". Our goal is to get this to work on Linux in our CI first and then make sure it works on Windows & Mac.

File: tools/wptrunner/wptrunner/executors/executorcontentshell.py
Comment (t-troebst) [reply tools/wptrunner/wptrunner/executors/executorcontentshell.py:51]: Also, just to clarify, the reason why ContentShellBrowser manages the stdio queues and threads is because of the architecture of wptrunner. ContentShellProtocol and ContentShellBrowser are running in different processes. The queues are not just there to provide non-blocking IO but to allow access to the IO from across processes in a clean way that works on all platforms. Sharing the handles themselves requires a bunch of platform-dependent voodoo AFAIK. Leaking the line-based nature of the protocol into the browser class is well worth it to not deal with that imo.

**Review by jcscottiii — APPROVED** [review]
LGTM
I would wait for @jonathan-j-lee's approval to see if their suggestions have been met.

**Review by jonathan-j-lee — APPROVED** [review]
LGTM. Having addressed the `ContentShellProtocol` API, I think this change should be good to land.

**Pull request merged**
